src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3012,24 +2984,61 @@ void G1CollectedHeap::do_collection_pause_at_safepoint_helper(double target_paus
   // We also do not allow mixed GCs during marking.
   assert(!collector_state()->mark_or_rebuild_in_progress() || collector_state()->in_young_only_phase(), "sanity");
 
+  return collector_state()->in_concurrent_start_gc();
+}
+
+void G1CollectedHeap::set_default_active_worker_threads(){
+  uint active_workers = WorkerPolicy::calc_active_workers(workers()->total_workers(),
+          workers()->active_workers(),
+          Threads::number_of_non_daemon_threads());
+  active_workers = workers()->update_active_workers(active_workers);
+  log_info(gc,task)("Using %u workers of %u for evacuation", active_workers, workers()->total_workers());
+}
+
+void G1CollectedHeap::prepare_tlabs_for_mutator() {
+  Ticks start = Ticks::now();
+
+  _survivor_evac_stats.adjust_desired_plab_sz();
+  _old_evac_stats.adjust_desired_plab_sz();
+
+  allocate_dummy_regions();
+
+  _allocator->init_mutator_alloc_regions();
+
+  resize_all_tlabs();
+
+  phase_times()->record_resize_tlab_time_ms((Ticks::now() - start).seconds() * 1000.0);
+}
+
+void G1CollectedHeap::retire_tlabs() {
+  ensure_parsability(true);
+}
+
+void G1CollectedHeap::do_collection_pause_at_safepoint_helper(double target_pause_time_ms) {
+  ResourceMark rm;
+
+  IsGCActiveMark active_gc_mark;
```

Comment:
Moving the IsGCActiveMark here expands the scope of is_gc_active() being true to cover areas it didn't used to.  I _think_ this is okay, but I didn't check all callers carefully.

Reply:
None of the code before that (verification, timing setup, heap printing) should be dependent on that. Actually I was considering moving the mark right after the guarantee in `G1CollectedHeap::do_collection_pause_at_safepoint`, but then I reconsidered due to the `GCLocker` check there.